python/_brotli.c

```diff
@@ -22,6 +22,32 @@
 #error "Only Python 3.10+ is supported"
 #endif
 
+/*
+Decoder / encoder nature does not support concurrent access. Attempt to enter
+concurrently will result in an exception.
+
+"Critical" parts used in prologues to ensure that only one thread enters.
+For consistency, we use them in epilogues as well. "Critical" is essential for
+free-threaded support and should be no-op in GIL environment (since methods are
+entered with GIL already held).
+
+"Native" parts denote code that does not interact with Python and thus GIL could
+be released. "Native" are important for GIL environment (for performance), but
+should be no-op in free-threaded environment, since prologue/epilogue should
+guarantee exclusive access.
+*/
+#ifdef Py_GIL_DISABLED
+#define BROTLI_CRITICAL_START Py_BEGIN_CRITICAL_SECTION(self)
+#define BROTLI_CRITICAL_END Py_END_CRITICAL_SECTION()
+#define BROTLI_NATIVE_START
+#define BROTLI_NATIVE_END
```

Comment:
These being no-ops is definitely wrong. You still have to explicitly detach from the interpreter before doing long-running operations on the free-threaded build.

This is the number one source of confusion when people think about updating extension modules. I wrote this to try to give you a mental model for how you should think about working with the C API:

https://py-free-threading.github.io/porting-extensions/#working-with-the-free-threaded-cpython-interpreter-runtime

ping @colesbury in case you want to weigh in as well